test/libsolidity/SolidityParser.cpp

```diff
@@ -1698,6 +1698,46 @@ BOOST_AUTO_TEST_CASE(newInvalidTypeName)
 	CHECK_PARSE_ERROR(text, "Expected explicit type name");
 }
 
+BOOST_AUTO_TEST_CASE(valid_function_type_variables)
+{
+	char const* text = R"(
+		contract test {
+			function(bytes memory) a;
+			function(bytes memory) internal b; // (explicit internal applies to the function type)
+			function(bytes memory) internal internal c;
+			function(bytes memory) external d;
+			function(bytes memory) external internal e;
+			function(bytes memory) internal public f;
+			function(bytes memory) internal pure public g;
+			function(bytes memory) pure internal public h;
```

Comment:
I think the "`FunctionType` type" could be verified with assigning a function to it, e.g.
```
function f(bytes memory) external {
}

function(bytes memory) a = f; // this should fail with an error
```

I know it may be a lot of extra test cases, but this is a syntactical element which I wouldn't call the best we have (in the future we should aim to clear this syntactical mess).